### ISSUE TITLE
ActiveSupport::BacktraceCleaner might wrongly remove supposed gem path

### ISSUE THREAD

**Issue by voxik** [rails/rails#40196]
It seems that the `ActiveSupport::BacktraceCleaner` is possibly removing parts of the backtrace paths, which resembles gem path but they are actually not gem path.
E.g. given that:
* the `gem_paths` RegExp [[1]] is `(?-mix:(\/builddir\/\.gem\/ruby|\/usr\/share\/gems|\/usr\/local\/share\/gems)\/(bundler\/)?gems\/([^\/]+)-([\w.]+)\/(.*))`
* and there is this backtrace line `/builddir/build/BUILD/web-console-4.0.4/usr/share/gems/gems/web-console-4.0.4/lib/web_console/evaluator.rb:28:in `eval'`,
* the output of such filter will be ```/builddir/build/BUILD/web-console-4.0.4web-console (4.0.4) lib/web_console/evaluator.rb:28:in `eval'``` which is wrong.

The possible fix might check for additional whitespace befor the path, e.g.:

~~~
$ git diff
diff --git a/activesupport/lib/active_support/backtrace_cleaner.rb b/activesupport/lib/active_support/backtrace_cleaner.rb
index ad80b8855f..785af44ba0 100644
--- a/activesupport/lib/active_support/backtrace_cleaner.rb
+++ b/activesupport/lib/active_support/backtrace_cleaner.rb
@@ -91,7 +91,7 @@ def add_gem_filter
         gems_paths = (Gem.path | [Gem.default_dir]).map { |p| Regexp.escape(p) }
         return if gems_paths.empty?
 
-        gems_regexp = %r{(#{gems_paths.join('|')})/(bundler/)?gems/([^/]+)-([\w.]+)/(.*)}
+        gems_regexp = %r{(#{gems_paths.map {|path| "\s#{path}"}.join('|')})/(bundler/)?gems/([^/]+)-([\w.]+)/(.*)}
         gems_result = '\3 (\4) \5'
         add_filter { |line| line.sub(gems_regexp, gems_result) }
       end
~~~

[1]: https://github.com/rails/rails/blob/580eefe1bc5d7241a4b2f3ee863523c5b8fd2f80/activesupport/lib/active_support/backtrace_cleaner.rb#L94

**Comment by composerinteralia** [comment]
Thanks for the issue! When you say "might wrongly remove supposed gem path", does that mean this is a hypothetical, or is this a bug you have seen in a real application? If you saw it in a real application, is there a way to reproduce the bug?

The linked [web-console issue](https://github.com/rails/web-console/issues/301) doesn't look related to me, at first glance. That seems to be a difference in how Ruby HEAD is generating the backtraces, and I am not seeing how this change would fix that.

**Comment by voxik** [comment]
I have tried update web-console package for Fedora, where during packaging we typically recreate rubygems directory structure nested under some buildroot. In the case above, the `/builddir/build/BUILD/web-console-4.0.4` is the buildroot and the `/usr/share/gems/gems/web-console-4.0.4/lib/web_console/evaluator.rb` is the nested Rubygems directory structure, which precisely corresponds with the final destination of web_console.

I have tried the above proposed patch in my scenario and it definitely helped. However, I am afraid that fixing this on Rails side is just one piece of the puzzle.
